drafttopic/utilities/fetch_wikiprojects.py

```diff
@@ -112,114 +111,95 @@ def parseWpDirectory(self):
+                wp[sec['line']]['topics'], _ = self.getSubCategories(
+                                                                  wp[sec['line']]['url'],
+                                                                  subPageSections,
+                                                                  0, 0)
         self.logger.info("Ended WikiProjects parsing")
         return wp
 
-    def getSubCategories(self, session, page):
-        """
-        Parses a one level down-sub directory page, looking for sections
-        and table listings
-        """
-        self.logger.info("Starting subdirectory {} parsing".format(page))
+    def getSubCategories(self, page, sections, index, level):
         wp = {}
-        sections = self.getSections(session, page)
-        titlestack = [sections[0]['line']]
-        prevLevel = sections[0]['toclevel']
-        wp['name'] = sections[0]['line']
-        wp['root_url'] = sections[0]['fromtitle']
-        wp['index'] = sections[0]['index']
-        wp['topics'] = {}
-        wpS = {}
-        wpS['topics'] = {}
-        wpS['topics'][wp['name']] = wp
-        for sec in sections:
-            # New level starts
-            if sec['toclevel'] > prevLevel:
-                initVals =  {'name': sec['line'], 'index':
-                            sec['index'], 'root_url':
-                            sec['fromtitle'], 'topics': {}}
-                titlestack.append(sec['line'])
-                tmpwp = wpS
-                for title in titlestack[:-1]:
-                    tmpwp = tmpwp['topics'][title]
-
-                tmpwp['topics'][sec['line']] = initVals
-            elif sec['toclevel'] < prevLevel:
-                titlestack.pop()
-                titlestack.pop()
-                initVals =  {'name': sec['line'], 'index':
-                            sec['index'], 'root_url':
-                            sec['fromtitle'], 'topics': {}}
-                titlestack.append(sec['line'])
-                tmpwp = wpS
-                for title in titlestack[:-1]:
-                    tmpwp = tmpwp['topics'][title]
-
-                tmpwp['topics'][sec['line']] = initVals
-            introProjects = self.getSectionIntro(session, page, sec['index'])
-            if introProjects:
-                self.setWikiProjects(wpS, titlestack, introProjects)
-            prevLevel = sec['toclevel']
-        return wpS['topics']
-
-    def setWikiProjects(self, wp, titlestack, projects, key = None):
-        tmpwp = wp
-        for title in titlestack:
-            tmpwp = tmpwp['topics'][title]
-        if key:
-            tmpwp['topics'][key] = {**tmpwp['topics'], **projects}
-        else:
-            tmpwp['topics'] = {**tmpwp['topics'], **projects}
-
-
-    def getSectionIntro(self, session, page, index):
+        prevTopic = None
+        self.logger.info("Index:{}, Level:{}".format(index, level))
+        idx = index
+        while idx < len(sections):
```

Comment:
would `for section in sections` be better here?

You can also do something like:
```
for i, section in enumerate(sections):
    ...
```

Reply:
I would've have done that but the recursion inherently depends on fast-forwarding the loop variable over parsed sections while exiting to avoid repeated calls to parse sections.